### PR TITLE
feat: schema for apisixroute

### PR DESCRIPTION
**Pull request by tokers**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues

___
### Bugfix
- Description

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

___
### Backport patches
- Why need to backport?

- Source branch

- Related commits and pull requests

- Target branch

### CODE REVIEW

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml [hunk 1]
```diff
@@ -35,3 +48,146 @@ spec:
...
+                        minItems: 1
+                        items:
+                          type: object
+                          properties:
+                            subject:
+                              type: object
+                              properties:
+                                scope:
+                                  type: string
+                                  enum: ["Header", "Query", "Cookie", "Path"]
```
Comment (jbampton) [inline samples/deploy/crd/v1beta1/ApisixRoute.yaml:112]: Should enums be sorted consistently the same way in each file ?

```suggestion
                                  enum: ["Cookie", "Header", "Path", "Query"]
```

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml [hunk 2]
```diff
@@ -35,3 +48,146 @@ spec:
...
+                        minItems: 1
+                        items:
+                          type: string
+                          pattern: "^\\*?[0-9a-zA-Z-._]+$"
+                      methods:
+                        type: array
+                        minItems: 1
+                        items:
+                          type: string
+                          enum: ["GET", "POST", "PUT", "DELETE", "PATCH", "HEAD", "OPTIONS", "CONNECT", "TRACE"]
```
Comment (jbampton) [inline samples/deploy/crd/v1beta1/ApisixRoute.yaml:95]: Whilst "GET" and "POST" are probably the most used or probably the most well known, should an enum array or list have a sort order or be random ?

```suggestion
                          enum: ["CONNECT", "DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT", "TRACE"]
```

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml [hunk 3]
```diff
@@ -35,3 +48,146 @@ spec:
...
+                                  type: string
+                                  enum: ["Header", "Query", "Cookie", "Path"]
+                                name:
+                                  type: string
+                                  minLength: 1
+                              required:
+                              - scope
+                            op:
+                              type: string
+                              enum:
```
Comment (jbampton) [inline samples/deploy/crd/v1beta1/ApisixRoute.yaml:120]: Should we sort enums ?  Or at least sort some part of the enums ?

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml [hunk 4]
```diff
@@ -35,3 +48,146 @@ spec:
...
+                                - Equal
+                                - NotEqual
+                                - GreaterThan
+                                - LessThan
+                                - In
+                                - NotIn
+                                - RegexMatch
```
Comment (jbampton) [inline samples/deploy/crd/v1beta1/ApisixRoute.yaml:127]: So for this enum maybe keep the top part with the custom order and then for the bottom part sort it ?

```suggestion
- RegexMatch
- RegexMatchCaseInsensitive
- RegexNotMatch
- RegexNotMatchCaseInsensitive
```

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml [hunk 5]
```diff
@@ -35,3 +48,146 @@ spec:
...
+                                - RegexMatchCaseInsensitive
+                                - RegexNotMatchCaseInsensitive
+                            value:
+                              type: string
+                            set:
+                              type: array
+                              items:
+                                type: string
+                          oneOf:
+                            - required: ["subject", "op", "value"]
```
Comment (jbampton) [inline samples/deploy/crd/v1beta1/ApisixRoute.yaml:138]:
```suggestion
                            - required: ["op", "subject", "value"]
```

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml [hunk 6]
```diff
@@ -35,3 +48,146 @@ spec:
...
+                            - required: ["subject", "op", "set"]
```
Comment (jbampton) [inline samples/deploy/crd/v1beta1/ApisixRoute.yaml:139]:
```suggestion
                            - required: ["op", "set", "subject"]
```

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml [hunk 7]
```diff
@@ -35,3 +48,146 @@ spec:
...
+        spec:
+          type: object
+          properties:
+            http:
+              type: array
+              minItems: 1
+              items:
+                type: object
+                oneOf:
+                  - required: ["name", "match", "backend"]
```
Comment (jbampton) [inline samples/deploy/crd/v1beta1/ApisixRoute.yaml:65]:
```suggestion
                  - required: ["backend", "match", "name"]
```

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml [hunk 8]
```diff
@@ -35,3 +48,146 @@ spec:
...
+                  - required: ["name", "match", "backends"]
```
Comment (jbampton) [inline samples/deploy/crd/v1beta1/ApisixRoute.yaml:66]:
```suggestion
                  - required: ["backends", "match", "name"]
```

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml (hunk 1 above)
Comment (tokers) [reply samples/deploy/crd/v1beta1/ApisixRoute.yaml:112]: Good point.

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml (hunk 4 above)
Comment (tokers) [reply samples/deploy/crd/v1beta1/ApisixRoute.yaml:127]: I grouped them by their meanings.

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml (hunk 7 above)
Comment (tokers) [reply samples/deploy/crd/v1beta1/ApisixRoute.yaml:65]: I grouped them by their meanings.

**Comment by tokers** [comment]
@jbampton Fields like `expr` operators can be grouped together by their meanings, the lexicographical orders are not so important.

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml [hunk 9]
```diff
@@ -35,3 +48,146 @@ spec:
...
+                        minItems: 1
+                        items:
+                          type: string
+                          pattern: "^/" # start with "/"
+                      hosts:
+                        type: array
+                        minItems: 1
+                        items:
+                          type: string
+                          pattern: "^\\*?[0-9a-zA-Z-._]+$"
```
Comment (gxthrj) [inline samples/deploy/crd/v1beta1/ApisixRoute.yaml:89]:
```suggestion
                          pattern: "(^\*?[a-zA-Z0-9._-]+$|^\*$)"
```

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml [hunk 10]
```diff
@@ -35,3 +48,146 @@ spec:
...
+                    type: object
+                    required:
+                    - paths
+                    properties:
+                      paths:
+                        type: array
+                        minItems: 1
+                        items:
+                          type: string
+                          pattern: "^/" # start with "/"
```
Comment (gxthrj) [inline samples/deploy/crd/v1beta1/ApisixRoute.yaml:83]:
```suggestion
                          pattern: "^\/[a-zA-Z0-9\-._~%!$&'()+,;=:@/]*\*?$"
```

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml (hunk 9 above)
Comment (tokers) [reply samples/deploy/crd/v1beta1/ApisixRoute.yaml:89]: I refer to APISIX.

**Review by gxthrj — APPROVED** [review]

**Pull request merged**